File: KmmTestApp/composeApp/src/commonMain/kotlin/com/gipstech/kmmtestapp/MyLocationListener.kt
package com.gipstech.kmmtestapp

import com.gipstech.mdk.CalibrationManager
import com.gipstech.mdk.CalibrationType
import com.gipstech.mdk.GiPStechException
import com.gipstech.mdk.Location
import com.gipstech.mdk.LocationListener
import com.gipstech.mdk.spatial.Building
import com.gipstech.mdk.spatial.Region

class MyLocationListener: LocationListener {
    private var currentBuilding: Building? = null

    override suspend fun onCalibrationRequest(types: Array<CalibrationType>) {
        LocationManager.locationMessage.value = "Calibration required!\nPlease rotate the phone to start the calibration process."
        CalibrationManager.beginCalibration(types[0], MyCalibrationListener(this))
    }

    override suspend fun onException(exception: GiPStechException) {
        LocationManager.locationMessage.value = "Error: ${exception.message}"
    }

    override suspend fun onBuildingChanged(building: Building?) {
        currentBuilding = building
        if (building != null) {
            LocationManager.levelMessage.value = "Building: ${building.name}"
        } else {
            LocationManager.levelMessage.value = "OUTDOOR"
        }
    }

    override suspend fun onLevelChanged(level: Int) {
        LocationManager.levelMessage.value = currentBuilding?.let {
             "Building: ${it.name}, Level: ${it.floors[level].name}"
        } ?: "OUTDOOR, Level: $level"
    }

    override suspend fun onLocationUpdated(location: Location) {
        LocationManager.locationMessage.value = "Location: ${location.longitude} ${location.latitude} [${location.steps}]"

        (LocationManager.place as? Region)?.let { region ->
            if (region.isNearOrOutsideBorder(location, 100.0)) {
                val (sw, ne) = location.getBoundingBox(1000.0, 1000.0)
                region.requestAdditionalArea(sw, ne)
            }
        }
    }

    override suspend fun onGeofenceEnter(name: String, confidence: Float) {
        LocationManager.geofenceMessage.value = "Enter: $name"
    }

    override suspend fun onGeofenceExit(name: String, confidence: Float) {
        LocationManager.geofenceMessage.value = "Exit: $name"
    }
}

File: SwiftTestApp/SwiftTestApp/MyLocationListener.swift
//
//  MyLocationListener.swift
//  SwiftTestApp
//
//  Created by GiPStech on 25/03/25.
//

import SwiftUI
import GiPStechMDK

class MyLocationListener: LocationListener {
    var currentBuilding: Building?
    var manager: LocationManager
    
    init(manager: LocationManager) {
        self.manager = manager
    }
    
    func __onBuildingChanged(building: Building?) async throws {
        currentBuilding = building
        manager.reportLevel("Building: \(currentBuilding?.name ?? "OUTDOOR")")
    }
    
    func __onCalibrationRequest(types: KotlinArray<__CalibrationType>) async throws {
        manager.report("Calibration required!\nPlease rotate the phone to start the calibration process.")
        try await CalibrationManager.shared.beginCalibration(
            type: types.get(index: 0)! as CalibrationType,
            listener: MyCalibrationListener(manager: manager))
    }
    
    func __onException(exception: GiPStechException) async throws {
        manager.report("Error: \(exception.message ?? "Unknown error")")
    }
    
    func __onGeofenceEnter(name: String, confidence: Float) async throws {
        manager.reportGeofence("Enter: \(name)")
    }
    
    func __onGeofenceExit(name: String, confidence: Float) async throws {
        manager.reportGeofence("Exit: \(name)")
    }
    
    func __onLevelChanged(level: Int32) async throws {
        if let building = currentBuilding {
            let floor = building.floors.get(index: level)
            manager.reportLevel("Building: \(building.name), Level: \(floor!.name)")
        } else {
            manager.reportLevel("OUTDOOR, Level: \(level)")
        }
    }
    
    func __onLocationUpdated(location: Location) async throws {
        if let region = manager.session?.target as? Region {
            if region.isNearOrOutsideBorder(coordinates: location, distance: 100.0) {
                let pair = location.getBoundingBox(size: 100)
                try await region.requestAdditionalArea(southWest: pair.first!, northEast: pair.second!, fallback: true)
            }
        }
        
        manager.report("Location: \(location.latitude) \(location.longitude) [\(location.steps)]")
    }
}
